docs/adr/adr-006-befp.md

```diff
@@ -0,0 +1,118 @@
+# ADR #006: Celestia-Node Bad encoding fraud proof(BEFP)
+
+## Changelog
+
+- 2022.03.03 - init commit
+- 2022.03.08 - added pub-sub
+
+## Authors
+
+@vgonkivs @Bidon15 @adlerjohn @Wondertan
+
+## Context
+
+In case when a Full Node receives `ErrByzantineRow`/`ErrByzantineCol` from the [rsmt2d](https://github.com/celestiaorg/rsmt2d) library, it generates a fraud proof.  After generation of it(fraud proof), a Full Node is broadcasting the fraud proof to Light Nodes, so they(Light Nodes) are notified that a block could be malicious.
+
+## Decision
+
+Started disscussion within:
+
+- https://github.com/celestiaorg/celestia-node/issues/4
+- https://github.com/celestiaorg/celestia-node/issues/263
+
+## Detailed Design
+A fraud proof is generated if recovered data does not match with its respective row/column roots during block repair process
+
+The result of RepairExtendedDataSquare will be an error [ErrByzantineRow](https://github.com/celestiaorg/rsmt2d/blob/f34ec414859fc834835ea97ed54300404eec1ac5/extendeddatacrossword.go#L18-L22)/[ErrByzantineCol](https://github.com/celestiaorg/rsmt2d/blob/f34ec414859fc834835ea97ed54300404eec1ac5/extendeddatacrossword.go#L28-L32):
+
+- Both errors consist of 
+  - row/column numbers that do not match with the Merkle root
+  - shares that were successfully repaired and verified (all correct shares).
+
+`ErrByzantineRow`/`ErrByzantineCol` are returned from `RetrevieData`. In addition, `das.Daser`:
+
+1. Generates a MerkleProofs for respective verified shares
+2. Creates a BEFP
+3. Broadcasts it(BEFP) to light nodes via separate sub-service.
+
+`das.Daser` imports a data structure that implements `proof.Broadcaster` interface that uses libp2p.pubsub under the hood:
+
+```go
+// Broadcaster is a generic interface that sends a different kinds of fraud proofs to all subscribed on particular topic nodes
+type Broadcaster interface {
+   // Broadcast takes a Fraud proof data stucture that implements standart BinaryMarshal interface and sends data to light nodes using libp2p pub-sub under the hood.
+   Broadcast(ctx context.Context, p encoding.BinaryMarshaller)  
+}
+```
```

Comment:
Is there a reason why this needs an interface? 

Will `Broadcaster` interface be implemented elsewhere outside of DASer? Is this done for testing purposes?

The name `Broadcaster` is confusing for this specific use case since `FraudSub` barely gets used. By barely gets used, I mean it is not a frequent event to broadcast a BEFP since this is a network-halting event.

Reply:
Do merkle proofs for correct shares get broadcasted on every block?

Reply:
> Do merkle proofs for correct shares get broadcasted on every block?

You mean under normal operation, orthogonal to BEFPs? Then...maybe? The sampling process involves descending through the tree, so by construction that means downloading a proof for each share sampled. Not sure if that's "broadcasted."

Reply:
As we discussed there are going to be multiple Fraud types that need to be broadcasted on the network, thus the interface. ALso, the msg is a bit outdated.